src/search/SearchMessagesScreen.js

```diff
@@ -131,6 +149,35 @@ class SearchMessagesScreenInner extends PureComponent<Props, State> {
     this.handleQuerySubmit(e);
   };
 
+  fetchOlder = async () => {
+    if (!this.props.loading && !this.state.foundOldest && !this.state.isFetching) {
+      this.setState({ isFetching: true });
+      try {
+        // Not sure why we require a double await here.
+        const { messages: fetchedMessages, foundOldest } = await await this.props.dispatch(
+          fetchMessages({
+            anchor: this.state.messages ? this.state.messages[0].id : LAST_MESSAGE_ANCHOR,
+            narrow: SEARCH_NARROW(this.state.query),
+            numBefore: config.messagesPerRequest,
+            numAfter: 0,
+          }),
+        );
+        // Last fetched message === first current message, hence remove it
+        fetchedMessages.pop();
+        this.setState(prevState => ({
+          messages: [...fetchedMessages, ...prevState.messages],
+          foundOldest,
```

Comment:
There's a tricky bug here. We're updating the state to add the messages we just fetched… but the query we fetched them for may no longer be the current query, the one that the other messages we're showing are from. That's because when we did the `await` above to wait for the results, the user may have entered a new query and those results may have come in. (Or for that matter a query they'd already entered, for which we had a request already in flight, may have had its results come in.)

A first attempt at fixing that would be to have the `setState` callback check that the current query is still the one that we did this fetch for, and if not then drop the update. So for example:
```js
    const { query } = this.state;
    const { messages: fetchedMessages, foundOldest } = await await this.props.dispatch(
        // …
        narrow: SEARCH_NARROW(query),
    // …
    this.setState(prevState => {
      if (prevState.query !== query) return {};
      return {
        messages: [...fetchedMessages, ...prevState.messages],
        foundOldest,
      };
```

I think that still isn't quite right, though, because it's possible that the query has changed and changed back, in which case the list of messages may no longer be the same -- so there might be a gap, or even an overlap, between the list of messages that's now in the state and the messages we've just fetched.

I think a sufficient fix would be to add a second condition, checking that the anchor message we fetched from is still the first message in the list in the current state: `prevState.messages && prevState.messages[0]?.id === anchor`, where `anchor` is a value computed before the `await` and passed to `fetchMessages`. (In addition to checking that the query is the same.)

An alternate approach would be to look at `this.lastIdSuccess` before the `await`, save that in a local variable, and check here that it's still the same. That would cover things completely, and it would make this logic mesh more with what's in `handleQuerySubmit`. But I kind of like having this *not* care about the implementation of `handleQuerySubmit` -- I think I like better the approach of checking `query` and `messages[0].id`, because I think that directly covers everything this fetch-older code actually cares about here.

Reply:
> There's a tricky bug here.
Indeed!
I have gone with the first approach, did test it using a fake delay and it does work out as far as I could tell.
```javascript
const sleep = ms => new Promise(
  resolve => setTimeout(resolve, ms));
``` 
Before, there were overlap between messages of different queries - if I changed queries in the middle of fetch, now that doesn't happen. A UX related problem that I didn't address is that the moment user presses the search button we should clear away the messages array and the loading spinner should again be visible, that is not the behavior we have currently.

Reply:
> A UX related problem that I didn't address is that the moment user presses the search button we should clear away the messages array and the loading spinner should again be visible, that is not the behavior we have currently.

Good point, yeah. Filed that as #5659 .

(As it sounds like you know, but just to make it explicit for anyone reading this quickly: that UX problem exists already in the current version of the app.)